src/java/org/jivesoftware/openfire/plugin/util/cache/ClusterListener.java

```diff
@@ -151,21 +153,49 @@ public void memberAdded(final MembershipEvent event) {
         final NodeID nodeID = ClusteredCacheFactory.getNodeID(event.getMember());
         if (event.getMember().localMember()) { // We left and re-joined the cluster
             joinCluster();
+
+            waitForClusterCacheToBeInstalled();
+
+            // Let the other nodes know that we joined the cluster
+            logger.debug("Done joining the cluster. Now proceed informing other nodes that we joined the cluster.");
+            CacheFactory.doClusterTask(new NewClusterMemberJoinedTask(true));
         } else {
-            if(wasSenior && !isSenior) {
+            if (wasSenior && !isSenior) {
                 logger.warn("Recovering from split-brain; firing leftCluster()/joinedCluster() events");
                 ClusteredCacheFactory.fireLeftClusterAndWaitToComplete(Duration.ofSeconds(30));
                 logger.debug("Firing joinedCluster() event");
                 ClusterManager.fireJoinedCluster(true);
-            } else {
-                // Trigger event that a new node has joined the cluster
-                ClusterManager.fireJoinedCluster(nodeID.toByteArray(), true);
+
+                waitForClusterCacheToBeInstalled();
+
+                // Let the other nodes know that we joined the cluster
+                logger.debug("Done joining the cluster in split brain recovery. Now proceed informing other nodes that we joined the cluster.");
+                CacheFactory.doClusterTask(new NewClusterMemberJoinedTask(true));
             }
         }
         clusterNodesInfo.put(nodeID,
                 new HazelcastClusterNodeInfo(event.getMember(), cluster.getClusterTime()));
     }
 
+    /**
+     * Blocks the current thread until the cluster cache is guaranteed to support clustering. This is especially useful
+     * for executing cluster tasks immediately after joining. If this wait is not performed, the cache factory may still
+     * be using the 'default' strategy instead of the 'hazelcast' strategy, which leads to cluster tasks being silently
+     * discarded.
+     */
+    private void waitForClusterCacheToBeInstalled() {
+        if (!ClusteredCacheFactory.PLUGIN_NAME.equals(CacheFactory.getPluginName())) {
+            logger.debug("This node now joined a cluster, but the cache factory has not been swapped to '{}' yet. Waiting for that to happen.", ClusteredCacheFactory.PLUGIN_NAME);
+            while (!ClusteredCacheFactory.PLUGIN_NAME.equals(CacheFactory.getPluginName())) {
+                try {
+                    Thread.sleep(200);
+                } catch (InterruptedException e) {
+                }
```

Comment:
Let's make sure that when the sleep is interrupted, the while-loop is broken. There's no other way to leave this loop otherwise (in case of failure / system shutdown etc). Also please log something, if only to debug. I've